tree/treeplayer/inc/ROOT/TVec.hxx

```diff
@@ -0,0 +1,381 @@
+}
+template <typename T, typename V>
+TVec<char> operator<=(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "<=", [](const T& t, const V &v) -> char {return t<=v;});
+}
+
+template <typename T, typename V>
+TVec<char> operator<(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "<", [](const T& t, const V &v) -> char {return t<v;});
+}
+
+// Workaround for operators >&co on vectors
+template <typename T>
+TVec<char> operator>(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator><T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator>=(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator>=<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator==(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator==<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator!=(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator!=<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator<=(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator<=<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator<(const TVec<T> &v0, const TVec<T> &v1)
+{
+   // clang-format off
+   return ROOT::Experimental::VecOps::operator< <T, T>(v0, v1);
+   // clang-format on
+}
+
+///@}
+
+/** @name Math Functions
+ *  Math functions on TVecs
+*/
+///@{
+#define MATH_FUNC(FUNCNAME)\
+template <typename T>\
+auto FUNCNAME(const TVec<T> &v) -> TVec<decltype(std::FUNCNAME(v[0]))>\
+{return ROOT::Internal::VecOps::Operate(v, [](const T& t) {return std::FUNCNAME(t);});}
+
+MATH_FUNC(sqrt)
+MATH_FUNC(log)
+MATH_FUNC(sin)
+MATH_FUNC(cos)
+MATH_FUNC(asin)
+MATH_FUNC(acos)
+MATH_FUNC(tan)
+MATH_FUNC(atan)
+MATH_FUNC(sinh)
+MATH_FUNC(cosh)
+MATH_FUNC(asinh)
+MATH_FUNC(acosh)
+MATH_FUNC(tanh)
+MATH_FUNC(atanh)
+
+
+///@}
+
+/// Inner product
+template <typename T, typename V>
+auto Dot(const TVec<T> v0, const TVec<V> v1) -> decltype(v0[0] * v1[0])
+{
+   ROOT::Internal::VecOps::CheckSizes(v0.size(), v1.size(), "Dot");
+   return std::inner_product(v0.begin(), v0.end(), v1.begin(), decltype(v0[0] * v1[0])(0));
+}
+
+/// Sum elements
+template <typename T>
+T Sum(const TVec<T> v)
+{
+   return std::accumulate(v.begin(), v.end(), 0);
+}
+
+template <class T>
+std::ostream &operator<<(std::ostream &os, const TVec<T> &v)
+{
+   // In order to print properly, convert to 64 bit int if this is a char
+   using Print_t = typename std::conditional<ROOT::Internal::VecOps::TIsChar<T>::value, long long int, T>::type;
+   os << "{ ";
+   auto size = v.size();
+   if (size) {
+      for (size_t i = 0; i < size - 1; ++i) {
+         os << (Print_t)v[i] << ", ";
+      }
+      os << (Print_t)v[size - 1];
+   }
+   os << " }";
+   return os;
+}
+
+template <typename T>
+TVec<T> Filter(const TVec<T> &v, bool cond)
```

Comment:
In what case would we need this?

Reply:
We support for all operations on TVecs the 2 cases: scalar and vector. This is done to preserve symmetry. A case which comes to my mind is filtering a collection, say jets, based on a scalar quantity, say MET or median energy density.

Reply:
To comment again, I am not sure how one could implement this behaviour in C++ once we buy into the syntax **v[v < 3]** since that would imply the "wrong" overload to be picked up :)

Reply:
This could be accomplished if `v < 3` returned some special type, like `TMaskVec<T>`, which would mean you always get the right overload. This is what is done in [`std::valarray`](http://en.cppreference.com/w/cpp/numeric/valarray/operator_at).